clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -633,56 +829,178 @@ public void assign(Collection<TopicPartition> partitions) {
-            final Deserializer<K> keyDeserializer,
-            final Deserializer<V> valueDeserializer,
-            final List<?>... candidateLists) {
-        ClusterResourceListeners clusterResourceListeners = new ClusterResourceListeners();
-        for (List<?> candidateList: candidateLists)
-            clusterResourceListeners.maybeAddAll(candidateList);
+    private Fetch<K, V> pollForFetches(Timer timer) {
+        long pollTimeout = timer.remainingMs();
+
+        // if data is available already, return it immediately
+        final Fetch<K, V> fetch = collectFetch();
+        if (!fetch.isEmpty()) {
+            return fetch;
+        }
+
+        // We do not want to be stuck blocking in poll if we are missing some positions
+        // since the offset lookup may be backing off after a failure
 
-        clusterResourceListeners.maybeAdd(keyDeserializer);
-        clusterResourceListeners.maybeAdd(valueDeserializer);
-        return clusterResourceListeners;
+        // NOTE: the use of cachedSubscriptionHasAllFetchPositions means we MUST call
+        // updateAssignmentMetadataIfNeeded before this method.
+        if (!cachedSubscriptionHasAllFetchPositions && pollTimeout > retryBackoffMs) {
+            pollTimeout = retryBackoffMs;
+        }
+
+        log.trace("Polling for fetches with timeout {}", pollTimeout);
+
+        Timer pollTimer = time.timer(pollTimeout);
+
+        // Wait a bit for some fetched data to arrive, as there may not be anything immediately available. Note the
+        // use of a shorter, dedicated "pollTimer" here which updates "timer" so that calling method (poll) will
+        // correctly handle the overall timeout.
+        try {
+            fetchBuffer.awaitNotEmpty(pollTimer);
+        } catch (InterruptException e) {
+            log.trace("Timeout during fetch", e);
+        } finally {
+            timer.update(pollTimer.currentTimeMs());
+        }
+
+        return collectFetch();
+    }
+
+    /**
+     * Perform the "{@link FetchCollector#collectFetch(FetchBuffer) fetch collection}" step by reading raw data out
+     * of the {@link #fetchBuffer}, converting it to a well-formed {@link CompletedFetch}, validating that it and
+     * the internal {@link SubscriptionState state} are correct, and then converting it all into a {@link Fetch}
+     * for returning.
+     *
+     * <p/>
+     *
+     * Note that if the {@link Fetch#isEmpty() is not empty}, this method will
+     * {@link ApplicationEventHandler#wakeup() wake up} the {@link ConsumerNetworkThread} before retuning. This is
+     * done as an optimization so that the <em>next round of data can be pre-fetched</em>.
+     */
+    private Fetch<K, V> collectFetch() {
+        final Fetch<K, V> fetch = fetchCollector.collectFetch(fetchBuffer);
+
+        if (!fetch.isEmpty()) {
+            // Notify the network thread to wake up and start the next round of fetching.
+            applicationEventHandler.wakeup();
```

Comment:
This doesn't seem quite right. It's possible for a fetchRequest to return no data. In that case, if we don't wake up the the network thread, it may not be able to send the next fetch request for a long time.

Reply:
The enqueuing of other events will also call `wakeupNetworkThread`, so it shouldn't be a problem. But I went ahead and changed it as requested.